src/rt/monitor_.d

```diff
@@ -1,444 +1,293 @@
+        auto m = cast(Monitor*) getMonitor(h);
+        assert(m.impl is null);
 
-        struct Monitor
+        foreach (ref v; m.devt)
         {
-            IMonitor impl; // for user-level monitors
-            DEvent[] devt; // for internal monitors
-            size_t   refs; // reference count
-            CRITICAL_SECTION mon;
+            if (v is null || v == e)
+            {
+                v = e;
+                return;
+            }
         }
+
+        auto len = m.devt.length + 4; // grow by 4 elements
+        auto pos = m.devt.length; // insert position
+        auto p = realloc(m.devt.ptr, DEvent.sizeof * len);
+        import core.exception : onOutOfMemoryError;
+
+        if (!p)
+            onOutOfMemoryError();
+        m.devt = (cast(DEvent*) p)[0 .. len];
+        m.devt[pos + 1 .. len] = null;
+        m.devt[pos] = e;
     }
-    else version( USE_PTHREADS )
+}
+
+extern (C) void rt_detachDisposeEvent(Object h, DEvent e)
+{
+    synchronized (h)
     {
-        import core.sys.posix.pthread;
+        auto m = cast(Monitor*) getMonitor(h);
+        assert(m.impl is null);
 
-        struct Monitor
+        foreach (p, v; m.devt)
         {
-            IMonitor impl; // for user-level monitors
-            DEvent[] devt; // for internal monitors
-            size_t   refs; // reference count
-            pthread_mutex_t mon;
+            if (v == e)
+            {
+                memmove(&m.devt[p], &m.devt[p + 1], (m.devt.length - p - 1) * DEvent.sizeof);
+                m.devt[$ - 1] = null;
+                return;
+            }
         }
     }
-    else
-    {
-        static assert(0, "Unsupported platform");
-    }
+}
 
-    Monitor* getMonitor(Object h) pure
-    {
-        return cast(Monitor*) h.__monitor;
-    }
+nothrow:
 
-    void setMonitor(Object h, Monitor* m) pure
+extern (C) void _d_monitor_staticctor()
+{
+    version (Posix)
     {
-        h.__monitor = m;
+        pthread_mutexattr_init(&gattr);
+        pthread_mutexattr_settype(&gattr, PTHREAD_MUTEX_RECURSIVE);
     }
+    initMutex(&gmtx);
+}
 
-    static __gshared int inited;
+extern (C) void _d_monitor_staticdtor()
+{
+    destroyMutex(&gmtx);
+    version (Posix)
+        pthread_mutexattr_destroy(&gattr);
 }
 
+package:
 
-/* =============================== Win32 ============================ */
+// This is what the monitor reference in Object points to
+alias IMonitor = Object.Monitor;
+alias DEvent = void delegate(Object);
 
-version( Windows )
+version (Windows)
 {
-    static __gshared CRITICAL_SECTION _monitor_critsec;
-
-    extern (C) void _STI_monitor_staticctor()
+    version (CRuntime_DigitalMars)
     {
-        debug(PRINTF) printf("+_STI_monitor_staticctor()\n");
-        if (!inited)
-        {
-            InitializeCriticalSection(&_monitor_critsec);
-            inited = 1;
-        }
-        debug(PRINTF) printf("-_STI_monitor_staticctor()\n");
+        pragma(lib, "snn.lib");
```

Comment:
As it says here, 759dc8d02fea6ef312a5cd24fd754300fb9e3b90, the C library isn't dragged in otherwise.